File: backend/java/src/main/java/walshe/projectcolumbo/persistence/service/SuperTrendCalculator.java
package walshe.projectcolumbo.persistence.service;
import walshe.projectcolumbo.persistence.model.SuperTrendResult;
import walshe.projectcolumbo.persistence.model.SuperTrendDirection;
import walshe.projectcolumbo.persistence.entity.Candle;

import org.springframework.stereotype.Component;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.List;

/**
 * Pure, stateless calculator for SuperTrend indicator.
 * Logic based on standard TradingView / SuperTrend implementation.
 */
@Component
public class SuperTrendCalculator {

    private static final int SCALE = 10;
    private static final RoundingMode ROUNDING = RoundingMode.HALF_UP;

    /**
     * Calculates SuperTrend for a list of candles.
     *
     * @param candles    Ordered list of candles (oldest to newest)
     * @param atrLength  ATR smoothing period (e.g., 10)
     * @param multiplier ATR multiplier (e.g., 3.0)
     * @return List of results, matching candle count (null values for initial period before ATR)
     */
    public List<SuperTrendResult> calculate(List<Candle> candles, int atrLength, BigDecimal multiplier) {
        List<SuperTrendResult> results = new ArrayList<>(candles.size());
        if (candles.isEmpty()) {
            return results;
        }

        BigDecimal[] tr = calculateTrueRanges(candles);
        BigDecimal[] atr = calculateATR(tr, atrLength);

        BigDecimal prevFinalUpper = null;
        BigDecimal prevFinalLower = null;
        BigDecimal prevSuperTrend = null;
        BigDecimal prevClose = null;

        for (int i = 0; i < candles.size(); i++) {
            Candle candle = candles.get(i);
            BigDecimal currentAtr = atr[i];

            if (currentAtr == null) {
                results.add(null);
                prevClose = candle.getClose();
                continue;
            }

            // 3. Basic Bands
            BigDecimal high = candle.getHigh();
            BigDecimal low = candle.getLow();
            BigDecimal close = candle.getClose();
            BigDecimal middle = high.add(low).divide(BigDecimal.valueOf(2), SCALE, ROUNDING);

            BigDecimal basicUpper = middle.add(multiplier.multiply(currentAtr));
            BigDecimal basicLower = middle.subtract(multiplier.multiply(currentAtr));

            // 4. Final Bands
            BigDecimal finalUpper;
            if (prevFinalUpper == null || basicUpper.compareTo(prevFinalUpper) < 0 || prevClose.compareTo(prevFinalUpper) > 0) {
                finalUpper = basicUpper;
            } else {
                finalUpper = prevFinalUpper;
            }

            BigDecimal finalLower;
            if (prevFinalLower == null || basicLower.compareTo(prevFinalLower) > 0 || prevClose.compareTo(prevFinalLower) < 0) {
                finalLower = basicLower;
            } else {
                finalLower = prevFinalLower;
            }

            // 5. SuperTrend
            BigDecimal superTrend;
            if (prevSuperTrend == null) {
                // Initial direction - assume DOWN if close <= finalUpper, else UP
                if (close.compareTo(finalUpper) <= 0) {
                    superTrend = finalUpper;
                } else {
                    superTrend = finalLower;
                }
            } else if (prevSuperTrend.equals(prevFinalUpper)) {
                if (close.compareTo(finalUpper) <= 0) {
                    superTrend = finalUpper;
                } else {
                    superTrend = finalLower;
                }
            } else { // prevSuperTrend == prevFinalLower
                if (close.compareTo(finalLower) >= 0) {
                    superTrend = finalLower;
                } else {
                    superTrend = finalUpper;
                }
            }

            // 6. Direction
            SuperTrendDirection direction = superTrend.equals(finalLower) ? SuperTrendDirection.UP : SuperTrendDirection.DOWN;

            results.add(new SuperTrendResult(
                    candle.getCloseTime(),
                    currentAtr,
                    finalUpper,
                    finalLower,
                    superTrend,
                    direction
            ));

            prevFinalUpper = finalUpper;
            prevFinalLower = finalLower;
            prevSuperTrend = superTrend;
            prevClose = close;
        }

        return results;
    }

    /**
     * Incremental calculation helper. Recomputes from a warm-up window around the last stored close time
     * to preserve ATR continuity and returns only the new results after the anchor.
     *
     * @param candles ordered list (oldest to newest)
     * @param atrLength ATR period
     * @param multiplier ATR multiplier
     * @param lastStoredCloseTime null if no prior data
     * @param fullRecalc when true, ignores incremental anchor and returns full results
     * @return list of new results strictly after {@code lastStoredCloseTime}; full list when {@code fullRecalc}
     */
    public List<SuperTrendResult> calculateIncremental(
            List<Candle> candles,
            int atrLength,
            BigDecimal multiplier,
            java.time.OffsetDateTime lastStoredCloseTime,
            boolean fullRecalc
    ) {
        if (fullRecalc || lastStoredCloseTime == null) {
            return calculate(candles, atrLength, multiplier);
        }
        if (candles.isEmpty()) {
            return List.of();
        }

        // Find anchor index: exact match, else first greater; if none greater → up to date
        int anchorIndex = -1;
        for (int i = 0; i < candles.size(); i++) {
            var ct = candles.get(i).getCloseTime();
            if (ct.equals(lastStoredCloseTime)) {
                anchorIndex = i;
                break;
            }
            if (ct.isAfter(lastStoredCloseTime)) {
                anchorIndex = i; // first greater
                break;
            }
        }
        if (anchorIndex == -1) {
            return List.of(); // already up to date
        }

        // Start from a warm-up window before the anchor point.
        // We use atrLength * 10 to ensure the ATR (Wilder's smoothing) has enough history to stabilize.
        int start = Math.max(0, anchorIndex - (atrLength * 10));
        List<Candle> window = candles.subList(start, candles.size());
        List<SuperTrendResult> windowResults = calculate(window, atrLength, multiplier);

        // Return only results with closeTime strictly after lastStoredCloseTime
        List<SuperTrendResult> out = new java.util.ArrayList<>();
        for (SuperTrendResult r : windowResults) {
            if (r == null) continue;
            if (r.closeTime().isAfter(lastStoredCloseTime)) {
                out.add(r);
            }
        }
        return out;
    }

    private BigDecimal[] calculateTrueRanges(List<Candle> candles) {
        BigDecimal[] tr = new BigDecimal[candles.size()];
        BigDecimal prevClose = null;

        for (int i = 0; i < candles.size(); i++) {
            Candle c = candles.get(i);
            BigDecimal highMinusLow = c.getHigh().subtract(c.getLow());

            if (prevClose == null) {
                tr[i] = highMinusLow;
            } else {
                BigDecimal highMinusPrevClose = c.getHigh().subtract(prevClose).abs();
                BigDecimal lowMinusPrevClose = c.getLow().subtract(prevClose).abs();
                tr[i] = highMinusLow.max(highMinusPrevClose).max(lowMinusPrevClose);
            }
            prevClose = c.getClose();
        }
        return tr;
    }

    private BigDecimal[] calculateATR(BigDecimal[] tr, int n) {
        BigDecimal[] atr = new BigDecimal[tr.length];
        if (tr.length < n) {
            return atr;
        }

        // First ATR = average of first n True Ranges
        BigDecimal sum = BigDecimal.ZERO;
        for (int i = 0; i < n; i++) {
            sum = sum.add(tr[i]);
        }
        atr[n - 1] = sum.divide(BigDecimal.valueOf(n), SCALE, ROUNDING);

        // Subsequent ATR = (previousATR * (n - 1) + currentTR) / n
        BigDecimal nMinusOne = BigDecimal.valueOf(n - 1);
        BigDecimal nDivisor = BigDecimal.valueOf(n);

        for (int i = n; i < tr.length; i++) {
            BigDecimal prevAtr = atr[i - 1];
            atr[i] = prevAtr.multiply(nMinusOne)
                    .add(tr[i])
                    .divide(nDivisor, SCALE, ROUNDING);
        }
        return atr;
    }
}

File: backend/supabase/functions/_shared/supertrend-calculator.ts
// Port of Java SuperTrendCalculator.java
// Pure, stateless calculator for SuperTrend indicator.
// Uses decimal.js-light for BigDecimal parity with Java (SCALE=10, HALF_UP).

import { Decimal, d, div, mul, SCALE } from "./decimal.ts";
import type { SuperTrendDirection, SuperTrendResult } from "./types.ts";

interface CandleInput {
  closeTime: string;
  high: string;
  low: string;
  close: string;
}

/**
 * Calculates SuperTrend for a list of candles.
 * Direct port of Java SuperTrendCalculator.calculate().
 *
 * @param candles Ordered list (oldest → newest)
 * @param atrLength ATR smoothing period (e.g. 10)
 * @param multiplier ATR multiplier (e.g. 2.0)
 * @returns List of results; null entries for initial period before ATR is available
 */
export function calculate(
  candles: CandleInput[],
  atrLength: number,
  multiplier: number | string
): (SuperTrendResult | null)[] {
  const results: (SuperTrendResult | null)[] = [];
  if (candles.length === 0) return results;

  const mult = d(multiplier);
  const tr = calculateTrueRanges(candles);
  const atr = calculateATR(tr, atrLength);

  let prevFinalUpper: Decimal | null = null;
  let prevFinalLower: Decimal | null = null;
  let prevSuperTrend: Decimal | null = null;
  let prevClose: Decimal | null = null;

  for (let i = 0; i < candles.length; i++) {
    const candle = candles[i];
    const currentAtr = atr[i];

    if (currentAtr === null) {
      results.push(null);
      prevClose = d(candle.close);
      continue;
    }

    // 3. Basic Bands
    const high = d(candle.high);
    const low = d(candle.low);
    const close = d(candle.close);
    const middle = div(high.plus(low), d(2));

    const basicUpper = middle.plus(mul(mult, currentAtr));
    const basicLower = middle.minus(mul(mult, currentAtr));

    // 4. Final Bands (sticky logic)
    let finalUpper: Decimal;
    if (
      prevFinalUpper === null ||
      basicUpper.lt(prevFinalUpper) ||
      prevClose!.gt(prevFinalUpper)
    ) {
      finalUpper = basicUpper;
    } else {
      finalUpper = prevFinalUpper;
    }

    let finalLower: Decimal;
    if (
      prevFinalLower === null ||
      basicLower.gt(prevFinalLower) ||
      prevClose!.lt(prevFinalLower)
    ) {
      finalLower = basicLower;
    } else {
      finalLower = prevFinalLower;
    }

    // 5. SuperTrend
    let superTrend: Decimal;
    if (prevSuperTrend === null) {
      // Initial direction: assume DOWN if close <= finalUpper, else UP
      superTrend = close.lte(finalUpper) ? finalUpper : finalLower;
    } else if (prevSuperTrend.eq(prevFinalUpper!)) {
      superTrend = close.lte(finalUpper) ? finalUpper : finalLower;
    } else {
      // prevSuperTrend == prevFinalLower
      superTrend = close.gte(finalLower) ? finalLower : finalUpper;
    }

    // 6. Direction
    const direction: SuperTrendDirection = superTrend.eq(finalLower)
      ? "UP"
      : "DOWN";

    results.push({
      closeTime: candle.closeTime,
      atr: currentAtr.toFixed(SCALE),
      upperBand: finalUpper.toFixed(SCALE),
      lowerBand: finalLower.toFixed(SCALE),
      supertrend: superTrend.toFixed(SCALE),
      direction,
    });

    prevFinalUpper = finalUpper;
    prevFinalLower = finalLower;
    prevSuperTrend = superTrend;
    prevClose = close;
  }

  return results;
}

/**
 * Incremental calculation. Recomputes from a warm-up window around the last stored
 * close time to preserve ATR continuity, and returns only new results after the anchor.
 * Direct port of Java SuperTrendCalculator.calculateIncremental().
 */
export function calculateIncremental(
  candles: CandleInput[],
  atrLength: number,
  multiplier: number | string,
  lastStoredCloseTime: string | null,
  fullRecalc: boolean
): SuperTrendResult[] {
  if (fullRecalc || lastStoredCloseTime === null) {
    return calculate(candles, atrLength, multiplier).filter(
      (r): r is SuperTrendResult => r !== null
    );
  }
  if (candles.length === 0) return [];

  // Find anchor index: exact match, else first greater; if none → up to date
  const lastStoredMs = new Date(lastStoredCloseTime).getTime();
  let anchorIndex = -1;
  for (let i = 0; i < candles.length; i++) {
    const ct = new Date(candles[i].closeTime).getTime();
    if (ct === lastStoredMs) {
      anchorIndex = i;
      break;
    }
    if (ct > lastStoredMs) {
      anchorIndex = i;
      break;
    }
  }
  if (anchorIndex === -1) return []; // already up to date

  // Start from a warm-up window before the anchor point.
  // atrLength * 10 ensures ATR (Wilder's smoothing) has enough history to stabilize.
  const start = Math.max(0, anchorIndex - atrLength * 10);
  const window = candles.slice(start);
  const windowResults = calculate(window, atrLength, multiplier);

  // Return only results with closeTime strictly after lastStoredCloseTime
  const out: SuperTrendResult[] = [];
  for (const r of windowResults) {
    if (r === null) continue;
    if (new Date(r.closeTime).getTime() > lastStoredMs) {
      out.push(r);
    }
  }
  return out;
}

// ── Internal helpers ────────────────────────────────────────────────────────

function calculateTrueRanges(candles: CandleInput[]): (Decimal | null)[] {
  const tr: (Decimal | null)[] = new Array(candles.length);
  let prevClose: Decimal | null = null;

  for (let i = 0; i < candles.length; i++) {
    const high = d(candles[i].high);
    const low = d(candles[i].low);
    const highMinusLow = high.minus(low);

    if (prevClose === null) {
      tr[i] = highMinusLow;
    } else {
      const highMinusPrevClose = high.minus(prevClose).abs();
      const lowMinusPrevClose = low.minus(prevClose).abs();
      tr[i] = Decimal.max(highMinusLow, highMinusPrevClose, lowMinusPrevClose);
    }
    prevClose = d(candles[i].close);
  }
  return tr;
}

function calculateATR(
  tr: (Decimal | null)[],
  n: number
): (Decimal | null)[] {
  const atr: (Decimal | null)[] = new Array(tr.length).fill(null);
  if (tr.length < n) return atr;

  // First ATR = simple average of first n True Ranges
  let sum = d(0);
  for (let i = 0; i < n; i++) {
    sum = sum.plus(tr[i]!);
  }
  atr[n - 1] = div(sum, d(n));

  // Subsequent ATR: Wilder smoothing = (prevATR * (n-1) + currentTR) / n
  const nMinusOne = d(n - 1);
  const nDivisor = d(n);

  for (let i = n; i < tr.length; i++) {
    const prevAtr = atr[i - 1]!;
    atr[i] = div(prevAtr.mul(nMinusOne).plus(tr[i]!), nDivisor);
  }
  return atr;
}
